### ISSUE TITLE
Missing dependency during unit testing

### ISSUE THREAD

**Issue by estruyf** [SharePoint/sp-dev-docs#526]
#### Category
- [x] Question
- [ ] Typo
- [x] Bug
- [ ] Enhancement / Suggestion
- [ ] Additional article idea

#### Expected or Desired Behavior

When I was writing unit tests against the webpart class, I get an error that the `@ms/sp-telemetry` is missing.

#### Observed Behavior

While running `gulp test` I receive the following error:

```
PhantomJS 2.1.1 (Mac OS X 0.0.0) ERROR
  Error: Cannot find module "@ms/sp-telemetry"
  at /Users/<user>/nodejs/spfx/testing-wp/temp/tests.js:52957
```

#### Steps to Reproduce

```typescript
describe('<HelloWorldWebPart />', () => {
    it('Should open the webpart', () => {
        const mockCtx = new MockWebPartContext();
        const wp = mockCtx.initialize(new HelloWorldWebPart());
    });
});
```
When I write this in a test I get the error. When I remove `const wp = mockCtx.initialize(new HelloWorldWebPart());`, all other tests keep working. 

When you just do something like this `const wp = new HelloWorldWebPart()` you get the same error.

**Comment by nickpape-msft** [comment]
@estruyf are you still seeing this issue?

**Comment by estruyf** [comment]
Yes, still the same missing dependency.

**Comment by estruyf** [comment]
Also the same in the new update.

**Comment by chan4lk** [comment]
Yes, still the same missing dependency.

**Comment by iclanton** [comment]
This is a known issue. We're working on a solution.

**Comment by Hennadiy** [comment]
I have the same error on test running if I import in the code:
import { HttpClient } from "@microsoft/sp-http";

**Comment by kbrimington** [comment]
Is there a workaround for this issue?

**Comment by bastianjohn** [comment]
We also have the same problem and need a workaround for this issue.

**Comment by tonywatters** [comment]
Have the same issue when running tests
Module not found: Error: Can't resolve '@ms/sp-telemetry'

any help ?

**Comment by buckyaustin** [comment]
Upgraded to latest sharepoint yoeman generator, and still getting this issue. I have removed node_modules folder, tried npm install / npm install --save-dev. Tried gulp clean. Module still missing.

Is there any update on this issue? Has anyone got around this issue and have a workaround?

**Comment by tonywatters** [comment]
I have run the following command to see packages
λ npm -g list --depth=0  
removed and updated phantomjs@2.1.7 and windows-build-tools@1.3.2
still having issues

**Comment by jacobhummel** [comment]
Also having issues with this.

**Comment by tonywatters** [comment]
Hi All, after some debugging and following @buckyaustin update.
I found that there was a duplicate @assets import within my WebPart, when I remove this the error "Module not found: Error: Can't resolve ms sp-telemetry" was gone and my unit tests started working again.

**Comment by deploymentvsureshk** [comment]
Not working me .. anu update

**Comment by FranckyC** [comment]
Same here. Any updates on this?

**Comment by michel-weber** [comment]
@iclanton Can you provide an update on this? 
Would be greatly appreciated :).

**Comment by deploymentvsureshk** [comment]
I checked and found that every method which is making a calll the http service needs to be stub
and it worked for me

**Comment by michel-weber** [comment]
@deploymentvsureshk Could be so kind as to create an example (HelloWorld WebPart) to demonstrate it?

**Comment by buckyaustin** [comment]
Further information that might help.

I have seen this occur on the following
@microsoft/sp-loader.SPComponentLoader.loadCss
@microsoft/sp-loader.SPComponentLoader.loadScript
@microsoft/sp-http.HttpClient
sp-pnp-js.pnp.web.lists

Just using the import statement for these doesn't cause the problem, but using methods within them causes the problem to occur.

I'm assuming there is a common class/method that all of these use that is causing the problem. 

You should stub responses from ajax or switch out a class that does the http call for another class that doesn't use http. In my build we switch classes so the http code doesn't get executed. Just by having the methods used in the code base is enough to cause the tests to fail even though they are not invoked in the tests. Commenting out the use of the methods allows gulp test to execute.

Hope this information helps.

**Comment by VelinGeorgiev** [comment]
Hey I am getting the same error as well. Version 1.2.0 of the SPFx. Seems that  "@ms/sp-telemetry"  package is included with almost all the SPFx libraries. When `gulp test` is run in the temp folder we get manifest and the `@ms/sp-telemetry` is hooked to almost every SPFx Library and Application including SPLoader, SPHttp, SPPageContext, SPComponentBase, SPWebPartBase, SPExtensionBase, WebPartWorkbench, SPLoader. However if we go to the SPFx node modules and reverse engeneer for a while we see that this is package is from internal Microsoft NPM server.
```
"description": "An internal package supporting the SharePoint Framework",
  "devDependencies": {
    "@ms/odsp-utilities-bundle": "~1.1.0",
    "@ms/sp-build-internal-web": "~0.6.1",
    "@ms/sp-telemetry": "~0.2.2",
    "@types/chai": "3.4.34",
    "@types/mocha": "2.2.38",
    "@types/sinon": "1.16.34",
    "@uifabric/styling": "0.15.2",
    "chai": "~3.5.0",
    "gulp": "~3.9.1",
    "sinon": "~1.17.6"
  }
```

Please remove that package from the loaderConfig so we can test or make it publicly available. Thanks!

**Comment by YonasJ** [comment]
This also affects "@microsoft/sp-dialog" where the internal  "@ms/sp-telemetry": "~0.2.2" is in the dev dependencies, But is used extensively in the code.

The missing class is a class to do some logging.

IMHO, It should be moved to the dependencies (from dev) for the module and made public, or all references to the logger should be removed, or a stub alternate provided.

My versions are:
```
"@microsoft/decorators": "~1.3.0",
    "@microsoft/sp-application-base": "1.3.0",
    "@microsoft/sp-core-library": "~1.3.0",
    "@microsoft/sp-dialog": "1.3.0",
    "@microsoft/sp-http": "^1.2.0",
    "@microsoft/sp-lodash-subset": "~1.3.0",
    "@microsoft/sp-webpart-base": "~1.3.0",
```

**Comment by VelinGeorgiev** [comment]
Another observation is that if you try for some reason to create/test new instance of class that uses `SPComponentLoader`  within .test.ts file then it will throw the `@ms/sp-telemetry` error and fail all tests. By removing such classes and replacing with mock-ed classes that do now have the reference to `SPComponentLoader`  might partially solve the issue.

**Comment by YonasJ** [comment]
I also get this error if I do this import:

`import { SPHttpClient, SPHttpClientResponse } from '@microsoft/sp-http';`

**Comment by ymihir** [comment]
I also faced same issue. But what i am having in my .ts file is
```
import SPHttpClient from '@microsoft/sp-http/lib/spHttpClient/SPHttpClient';
import SPHttpClientResponse from '@microsoft/sp-http/lib/spHttpClient/SPHttpClientResponse';
```
Since I am importing spHttpClient twice even though for different modules, it is giving error. If I change to `import {SPHttpClient,SPHttpClientResponse} from '@microsoft/sp-http';` then it started working.
Hope this is helpful.

**Comment by PavelAl** [comment]
@ymihir solution worked for me as well.

**Comment by smithywick** [comment]
With 1.3 this issue prevents testing of spHttpClient.

**Comment by tracy-zaact** [comment]
@VesaJuvonen  - I see this is among the 6 issues still marked as 'status:fixed-next-drop'.  Can someone tell me what that means from an ETA standpoint?

**Comment by AustinBreslinDev** [comment]
I can reproduce issue in spfx 1.4.0.

**Comment by lahuey** [comment]
Hi,
Thank you for the feedback. Please add a feature request on UserVoice and depending on community interest we may or may not be able to make the fix.

**Comment by tracy-zaact** [comment]
@lahuey - Please correct me if I'm mistaken, but I don't think this is a feature request - it's a bug.

**Comment by michel-weber** [comment]
Is this fixed in 1.4.1?

**Comment by jacobhummel** [comment]
@michel-weber We're still experiencing this issue in 1.4.1. 

I saw the tag _status:fixed-next-drop_ so I was under the impression that a fix would be rolled out soon. Is that not the case?

If this has been de-prioritized, does anyone have any docs/blogs explaining how to work around this? This issue is blocking my team from doing any testing on web parts which include REST calls (all of our components do). Thanks for any info.

**Comment by michel-weber** [comment]
That was also my understanding. But the tag was added on the 28th of October 2017 - and since then 1.4.0 and 1.4.1 have been released. So I think this issue should be reopened.

@VesaJuvonen Can you look into this? Will this fix be released with the next version (> 1.4.1)?

**Comment by VelinGeorgiev** [comment]
I think @lahuey is silently saying  "Deal with it" :) . It has always been convenient to turn bug into a feature :) . Considering the fact that the testing scaffold templates are not in the latest framework anymore and the last comment from @patmill here https://github.com/SharePoint/sp-dev-docs/issues/1335. I am not sure if we should still rely on the testing framework that comes with the frameworks anymore.

**Comment by kezack** [comment]
I have the same problem with import { GraphHttpClient, HttpClientResponse} from '@microsoft/sp-http';

Error: Cannot find module "@ms/sp-telemetry"

have you resolve this problem ?

**Comment by ericchin79** [comment]
We are running into the same issue here

Cannot find module '@ms/sp-telemetry' from 'DigestCache.js'
at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:179:17)
at Object.<anonymous> (node_modules/@microsoft/sp-http/lib/spHttpClient/DigestCache.js:3:22)

**Comment by devinprejean** [comment]
I have a possible workaround for this!
This worked for me, Your mileage may vary.

I had narrowed down my issue specifically to when i would "new up" an object that didn't even reference anything SP, HTTP, or graph client. I found that instead of newing an object, I casted a general one into my class and ran the tests worked.

I had this 
`const testObject = new WeatherInfo(params...)`
now i have this 
```
const tempObj = {
    param1: ...
    param2: ...
};
const realObjectToTestWith = testObject as WeatherInfo;

```

**Comment by devinprejean** [comment]
Sorry for the double post but I took another stab at my issue and found a fix for me, not a workaround.
Seems I was referencing imports from the same folder, however one group was listed in my index, the other was referenced all the way. For example:
```
import { ModuleName } From "./provider/module
import { SecondModule } from "./provider" 
```

once i changed the second module to reference all the way instead of using the index.ts file inside the folder my issue went away.
Hope this helps someone.

**Comment by deploymentvsureshk** [comment]
This is due to if you have any http reference used for TEST
    On Saturday, 7 July, 2018, 1:15:53 AM IST, Devin Prejean <notifications@github.com> wrote:  
 
 
Sorry for the double post but I took another stab at my issue and found a fix for me, not a workaround.
Seems I was referencing imports from the same folder, however one group was listed in my index, the other was referenced all the way. For example:
import { ModuleName } From "./provider/module
import { SecondModule } from "./provider" 

once i changed the second module to reference all the way instead of using the index.ts file inside the folder my issue went away.
Hope this helps someone.

**Comment by adambarrettsmith** [comment]
I'm running into a similar issue:

Cannot find module "@ms/sp-client-shared" which is being referenced under "..\node_modules\@microsoft\sp-http\lib\oauthTokenProvider\OAuthTokenProvider.js"

**Comment by samyaktelsang** [comment]
Hey @adambarrettsmith . I'm running into a similar issue. Wondering if you could update this thread if you found a workaround. Thanks!

**Comment by grantbowering** [comment]
For the record:
I was experiencing this on `1.4.1`, and ultimately ended up upgrading to `1.6.0-plusbeta` which seems to have eliminated the problem.  
(Recommendation: use `office365-cli` to retrieve steps for doing this upgrade, as there are a bunch of minor manual config changes it reports)

**Comment by rabwill** [comment]
I am getting this error ` Cannot find module '@ms/sp-client-shared' from 'OAuthTokenProvider.js` when I try to run tests for services that use `SPHttpClient` for rest calls. 
In my jest configuration I have 

```sh
"transformIgnorePatterns": [ "node_modules/(?!(@microsoft/sp-http)|(@microsoft/sp-core-library)|(@microsoft/sp-diagnostics))" ],

```
Any help or direction in this space would be helpful as I am new to writing tests.
I am using **Sinon** stubs to mock the functions.

**Comment by rvacchi** [comment]
Dear all I am running into this problem when I try to debug a spfx react web part Cannot find module '@ms/sp-telemetry'

Can someone help?

**Comment by kogermd** [comment]
Has this been resolved? I am on SPFx v1.8.0 and I am seeing missing dependencies when trying to unit test my web parts (which are not using React) using Jest. Specifically I am seeing _Cannot find module '@ms/sp-telemetry' from 'SPStarter.js'_ when trying to import SPComponentLoader from @microsoft/sp-loader in my web part. If I remove that from my web part, I then encounter another error: _Cannot find module '@ms/sp-client-shared' from 'OAuthTokenProvider.js'_.

Am I missing something? Thanks in advance!

**Comment by rabwill** [comment]
Running into the same issue for unit tests in SPFx using jest/sinon.
One of the react components use 
```
import { SPComponentLoader } from '@microsoft/sp-loader';
```
Tests fail with message
```
  ● Test suite failed to run

    Cannot find module '@ms/sp-telemetry' from 'SPStarter.js'

      at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
      at Object.<anonymous> (node_modules/@microsoft/sp-loader/lib/SPStarter.js:4:22)

```
SPFx version 1.8.2
jest : 24.8.0
npm : 6.9.0
node : 10.15.3
